src/main/kotlin/org/rust/ide/intentions/NestUseStatementsIntention.kt

```diff
@@ -0,0 +1,127 @@
+        val useSpecksByBasePath = mapSpecksByBasePath(ctx)
+
+        val createdUseItems = createGroupedUseItems(useSpecksByBasePath, project)
+
+        val root = ctx.firstUseItem.parent
+        var firstUseSpeck: PsiElement? = null
+        var lastUseSpeck: PsiElement? = null
+
+        for (createdUseItem in createdUseItems) {
+            val addPoint = lastUseSpeck ?: ctx.firstUseItem
+            val added = root.addAfter(createdUseItem, addPoint)
+            lastUseSpeck = added
+            firstUseSpeck = firstUseSpeck ?: added
+        }
+
+        for (useItem in ctx.useItems) {
+            useItem.delete()
+        }
+
+        editor.caretModel.moveToOffset(firstUseSpeck!!.textRange.startOffset + "use ".length)
+    }
+
+    private fun mapSpecksByBasePath(ctx: Context): MutableMap<String, MutableList<RsUseSpeck>> {
+        val useSpeckByBasePath = mutableMapOf<String, MutableList<RsUseSpeck>>()
+        for (useSpeck in ctx.useSpecks) {
+            val useSpeckStartsWithGroup = useSpeck.path == null
+            if (useSpeckStartsWithGroup) {
+                useSpeck.forEachLeafSpeck {
+                    useSpeckByBasePath.getOrPut(it.path!!.basePath().text) { mutableListOf() }.add(it)
+                }
+            } else {
+                useSpeckByBasePath.getOrPut(useSpeck.path!!.basePath().text) { mutableListOf() }.add(useSpeck)
+            }
+        }
+        return useSpeckByBasePath
+    }
+
+    private fun createGroupedUseItems(useSpecksByBasePath: MutableMap<String, MutableList<RsUseSpeck>>, project: Project): MutableList<RsUseItem> {
+        val createdUseItems = mutableListOf<RsUseItem>()
+        for (basePath in useSpecksByBasePath.keys) {
+            // Cannot group use specks with this $basePath
+            if (useSpecksByBasePath[basePath]!!.count() == 1) {
+                val useSpeck = useSpecksByBasePath[basePath]!![0]
+                val newUseItem = RsPsiFactory(project).createUseItem(useSpeck.text)
+                createdUseItems.add(newUseItem)
+                // Group use specks with this $basePath
+            } else {
+                val useSpecksInGroup = mutableListOf<String>()
+                for (eachUseSpeck in useSpecksByBasePath[basePath]!!) {
+                    val withoutBasePath = deleteBasePath(eachUseSpeck.text, basePath)
+                    useSpecksInGroup.add(withoutBasePath)
+                }
+
+                val groupString = "{\n${useSpecksInGroup.joinToString(",\n")},\n}"
+                val newUseItem = RsPsiFactory(project).createUseItem("$basePath::$groupString")
+                createdUseItems.add(newUseItem)
+            }
+        }
+        return createdUseItems
+    }
+
+    private fun deleteBasePath(fullPath: String, basePath: String): String {
+        return when {
+            fullPath.startsWith("::$basePath") -> {
+                fullPath.substring("::$basePath::".length)
```

Comment:
Kotlin's function `removePrefix` can be used. And maybe it allows to rewrite it with 2 lines of code